### PR TITLE
Change notification to use builder pattern.

### PR DESCRIPTION
**Pull request by terrynsun**

### CODE REVIEW

**Comment by terrynsun** [comment]
What does Notification.Builder mean? I added a new file and class NotificationBuilder, but I can update this.

**Comment by Topher-the-Geek** [comment]
Make the builder a nested class in the Notification object. That's a common
place to put the builder.

File: disk/test/com/treode/disk/DiskTestTools.scala [hunk 1]
```diff
@@ -40,6 +41,17 @@ private object DiskTestTools {
 
   val sysid = SystemId (0, 0)
 
+  def assertEqNotification (msgs: List[String], notif: Notification) {
```
Comment (Topher-the-Geek) [inline disk/test/com/treode/disk/DiskTestTools.scala]: If you define this as `def assertError (msgs: String*) (actual: Notification)`, you save the need for the `List(...)` part at the call sites.

**Comment by terrynsun** [comment]
Sorry about the mistakes. I didn't understand that you could define functions with two sets of arguments.

**Comment by Topher-the-Geek** [comment]
No worries. Learn by doing.

Now we're going to take asserEqNotification one step further.

`def assertEqNotification (msgs: Message*) (notif: Notification)`

Then your call sites can be things like:

`assertEqNotification (AlreadyAttaching ("f1")) (e)`

Now your test doesn't depend on the message text; we could rewrite the message text somday down the road without invalidating the tests. Also, now your tests hook into the constructor. If we added params to the message someday down the road, we'd get a compile error rather than a test error.

File: core/src/com/treode/notify/Notification.scala [hunk 2]
```diff
@@ -23,14 +23,27 @@ package com.treode.notify
   * Inspired by [[http://martinfowler.com/articles/replaceThrowWithNotification.html Martin Fowler]].
   */
 
-class Notification {
```
Comment (Topher-the-Geek) [inline core/src/com/treode/notify/Notification.scala]: I'd recommend giving Notification two abstract methods: isEmpty and messages, and fill in the bodies for Errors and NoErrors. Actually, the body for messages is already filled in for Errors.

File: disk/src/com/treode/disk/edit/DriveGroup.scala [hunk 3]
```diff
@@ -264,20 +265,21 @@ private class DriveGroup (
...
-        // Otherwise, we can merge our new changes into the queued changes.
-        this.dno = dno
-        attaches :::= newAttaches
-        drains :::= newDrains
-        changers ::= cb
-        queue.engage()
+      val errorlist = errors.result
+
+      errors.result match {
+        case Errors (list) =>
```
Comment (Topher-the-Geek) [inline disk/src/com/treode/disk/edit/DriveGroup.scala]: if you say `case notify @ Errors (_) =>` here, you can use notify for the callback below, and you can remove the `errorlist = errors.result` above

File: disk/test/com/treode/disk/DiskTestTools.scala [hunk 4]
```diff
@@ -40,6 +41,17 @@ private object DiskTestTools {
 
   val sysid = SystemId (0, 0)
 
+  def assertEqNotification (msgs: String*) (notif: Notification) {
+    notif match {
+      case NoErrors () => assert (msgs.length == 0)
```
Comment (Topher-the-Geek) [inline disk/test/com/treode/disk/DiskTestTools.scala]: With the abstract functions, you no longer need to pattern match here. You can just say `msgs == notify.messages`. That kind of weirdness, though, shows why `expected` and `actual` make good param names. It'd be `expected == actual.messages`

File: disk/test/com/treode/disk/DiskTestTools.scala [hunk 5]
```diff
@@ -40,6 +41,15 @@ private object DiskTestTools {
...
 
+  def assertEqNotification (expected: Message*) (actual: Notification) {
+    if (actual.isEmpty) {
+      assert (expected.length == 0)
+    } else {
+      assert (expected == actual.messages)
+      assert (expected.length == actual.messages.length)
+    }
+  }
+
```
Comment (Topher-the-Geek) [inline disk/test/com/treode/disk/DiskTestTools.scala]: Awesome. You're almost there. All that's left is to simplify this. `expected == actual.messages` will do. There's no need to special case empty. After checking that the sequences are equal, it's redundant to check that their lengths are equals.

**Pull request closed without merging**